packages/java/endpoint/src/main/java/com/vaadin/hilla/route/RouteExtractionIndexHtmlRequestListener.java

```diff
@@ -0,0 +1,130 @@
+package com.vaadin.hilla.route;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.vaadin.flow.router.PageTitle;
+import com.vaadin.flow.router.RouteData;
+import com.vaadin.flow.server.RouteRegistry;
+import com.vaadin.flow.server.VaadinService;
+import com.vaadin.flow.server.communication.IndexHtmlRequestListener;
+import com.vaadin.flow.server.communication.IndexHtmlResponse;
+import com.vaadin.hilla.route.records.AvailableViewInfo;
+import com.vaadin.hilla.route.records.RouteParamType;
+import org.jsoup.nodes.DataNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Index HTML request listener for collecting the client side and the server
+ * side views and adding them to index.html response.
+ */
+@Component
+public class RouteExtractionIndexHtmlRequestListener
+        implements IndexHtmlRequestListener {
+    protected static final String SCRIPT_STRING = "window.Vaadin = window.Vaadin ?? {}; "
+            + " window.Vaadin.server = window.Vaadin.server ?? {}; "
+            + " window.Vaadin.server.views = %s;";
+    private static final Logger LOGGER = LoggerFactory
+            .getLogger(RouteExtractionIndexHtmlRequestListener.class);
+    private final ObjectMapper mapper = new ObjectMapper();
+
+    private final ClientRouteRegistry clientRouteRegistry;
+
+    /**
+     * Creates a new listener instance with the given route registry.
+     *
+     * @param clientRouteRegistry
+     *            the client route registry for getting the client side views
+     */
+    @Autowired
+    public RouteExtractionIndexHtmlRequestListener(
+            ClientRouteRegistry clientRouteRegistry) {
+        this.clientRouteRegistry = clientRouteRegistry;
+    }
+
+    @Override
+    public void modifyIndexHtmlResponse(IndexHtmlResponse response) {
+        final List<AvailableViewInfo> availableViews = new ArrayList<>();
+        collectClientViews(availableViews);
+        collectServerViews(availableViews);
+
+        if (availableViews.isEmpty()) {
+            return;
+        }
+        try {
+            final String viewsJson = mapper.writeValueAsString(availableViews);
+            final String script = SCRIPT_STRING.formatted(viewsJson);
+            response.getDocument().head().appendElement("script")
+                    .appendChild(new DataNode(script));
+        } catch (IOException e) {
+            LOGGER.warn("Failed to write server views to index response", e);
```

Comment:
I wonder if this should be a runtime error that creates a BUZZ about the error in serializing the available views, or at least an ERROR instead of WARN?